src/main.rs

```diff
@@ -0,0 +1,487 @@
+        position += progress as u64;
+        if last_report.elapsed() >= Duration::from_secs(1)
+            || (have_length && position == length_hint)
+        {
+            last_report = Instant::now();
+            if have_length {
+                print!(
+                    "> Read {}/{} ({}%)   \r",
+                    format_bytes(position),
+                    format_bytes(length_hint),
+                    100 * position / length_hint
+                );
+            } else {
+                print!("> Read {}   \r", format_bytes(position));
+            }
+            let _ = std::io::stdout().flush();
+        }
+    });
+
+    // Wrap in a BufReader so we can peek at the first few bytes for
+    // format sniffing.  Don't trust the content-type since the server
+    // may not be configured correctly, or the file might be local.
+    // Then wrap in a reader for decompression.
+    let mut buf_reader = BufReader::new(&mut progress_reader);
+    let mut decompress_reader: Box<dyn Read> = {
+        let sniff = buf_reader.fill_buf().chain_err(|| "sniffing input")?;
+        // verify default buffer size >= the largest magic number we might
+        // care about
+        assert!(sniff.len() >= 8);
+        if &sniff[0..2] == b"\x1f\x8b" {
+            Box::new(GzDecoder::new(buf_reader))
+        } else if &sniff[0..6] == b"\xfd7zXZ\x00" {
+            Box::new(XzDecoder::new(buf_reader))
+        } else {
+            Box::new(buf_reader)
+        }
+    };
+
+    // Cache the first MiB of input and write zeroes instead.  This ensures
+    // that the disk image can't be used accidentally before its GPG signature
+    // is verified.
+    let mut first_mb: [u8; 1024 * 1024] = [0; 1024 * 1024];
+    dest.write_all(&first_mb)
+        .chain_err(|| "clearing first MiB of disk")?;
+    decompress_reader
+        .read_exact(&mut first_mb)
+        .chain_err(|| "decoding first MiB of image")?;
+
+    // do the rest of the copy
+    copy(&mut decompress_reader, dest).chain_err(|| "decoding and writing image")?;
+
+    // verify_reader has now checked the signature, so fill in the first MiB
+    dest.seek(SeekFrom::Start(0))
+        .chain_err(|| "seeking to start of disk")?;
+    dest.write_all(&first_mb)
+        .chain_err(|| "writing to first MiB of disk")?;
+
+    // flush
+    dest.flush().chain_err(|| "flushing data to disk")?;
+    dest.sync_all().chain_err(|| "syncing data to disk")?;
+
+    // log final newline
+    println!();
+
+    Ok(())
+}
+
+/// Write the Ignition config.
+fn write_ignition(mountpoint: &Path, config_src: &str) -> Result<()> {
+    println!("Writing Ignition config");
+
+    // make parent directory
+    let mut config_dest = mountpoint.to_path_buf();
+    config_dest.push("ignition");
+    create_dir_all(&config_dest).chain_err(|| "creating Ignition config directory")?;
+
+    // do the copy
+    config_dest.push("config.ign");
+    let mut config_in = OpenOptions::new()
```

Comment:
Why not just `std::fs::copy` here?

Reply:
We get a bit better control this way:

- `config_out.create_new(true)` ensures that we'll fail if the file already exists for some reason.  (Which it shouldn't, of course.)
- `std::fs::copy` also copies permissions, which may not be desired.